pkg/interface/src/views/apps/links/components/LinkSubmit.tsx

```diff
@@ -0,0 +1,230 @@
+import { BaseInput, Box, Button, LoadingSpinner, Text } from "@tlon/indigo-react";
+import React, { useCallback, useState } from "react";
+import GlobalApi from "~/logic/api/global";
+import { useFileDrag } from "~/logic/lib/useDrag";
+import useS3 from "~/logic/lib/useS3";
+import { S3State } from "~/types";
+import SubmitDragger from "~/views/components/SubmitDragger";
+import { createPost } from "~/logic/api/graph";
+import { hasProvider } from "oembed-parser";
+
+interface LinkSubmitProps {
+  api: GlobalApi;
+  s3: S3State;
+  name: string;
+  ship: string;
+};
+
+const LinkSubmit = (props: LinkSubmitProps) => {
+  let { canUpload, uploadDefault, uploading, promptUpload } = useS3(props.s3);
+
+  const [submitFocused, setSubmitFocused] = useState(false);
+  const [urlFocused, setUrlFocused] = useState(false);
+  const [linkValue, setLinkValueHook] = useState('');
+  const [linkTitle, setLinkTitle] = useState('');
+  const [disabled, setDisabled] = useState(false);
+  const [linkValid, setLinkValid] = useState(false);
+
+  const doPost = () => {
+    const url = linkValue;
+    const text = linkTitle ? linkTitle : linkValue;
+    setDisabled(true);
+    const parentIndex = props.parentIndex || '';
+    const post = createPost([
+      { text },
+      { url }
+    ], parentIndex);
+
+    props.api.graph.addPost(
+      `~${props.ship}`,
+      props.name,
+      post
+    ).then(() => {
+      setDisabled(false);
+      setLinkValue('');
+      setLinkTitle('');
+      setLinkValid(false);
+    });
+  };
+
+  const validateLink = (link) => {
+    const URLparser = new RegExp(
+      /((?:([\w\d\.-]+)\:\/\/?){1}(?:(www)\.?){0,1}(((?:[\w\d-]+\.)*)([\w\d-]+\.[\w\d]+))){1}(?:\:(\d+)){0,1}((\/(?:(?:[^\/\s\?]+\/)*))(?:([^\?\/\s#]+?(?:.[^\?\s]+){0,1}){0,1}(?:\?([^\s#]+)){0,1})){0,1}(?:#([^#\s]+)){0,1}/
+    );
+
+    let linkValid = URLparser.test(link);
+
+    if (!linkValid) {
+      linkValid = URLparser.test(`http://${link}`);
+      if (linkValid) {
+        link = `http://${link}`;
+        setLinkValue(link);
+      }
+    }
+
+    if (linkValid) {
+      if (hasProvider(linkValue)) {
+        fetch(`https://noembed.com/embed?url=${linkValue}`)
+          .then(response => response.json())
+          .then((result) => {
+            if (result.title && !linkTitle) {
+              setLinkTitle(result.title);
+            }
+          }).catch((error) => { /* noop*/ });
+      } else if (!linkTitle) {
+        setLinkTitle(decodeURIComponent(link
+          .split('/')
+          .pop()
+          .split('.')
+          .slice(0, -1)
+          .join('.')
+          .replace('_', ' ')
+          .replace(/\d{4}\.\d{1,2}\.\d{2}\.\.\d{2}\.\d{2}\.\d{2}-/, '')
+        ));
+      }
+    }
+    return link;
+  };
+
+  const onFileDrag = useCallback(
+    async (files: FileList | File[], e: DragEvent) => {
+      if (!canUpload) {
+        return;
+      }
+      uploadDefault(files[0]).then(setLinkValue);
```

Comment:
not returning or awating a promise?

Reply:
Not sure what you mean. This function doesn't need to return anything:
<img width="646" alt="image" src="https://user-images.githubusercontent.com/3999320/101058523-1288c900-3542-11eb-9ee5-e5ea98d9d44d.png">

Reply:
this callback is marked `async` so surely you'd use `await` instead of chaining promises? Either that or drop the `async`

Reply:
Good point. Must have been an oversight.